examples/generative/vq_vae.py

```diff
@@ -0,0 +1,563 @@
+outputs. These discrete code words are then fed to the decoder, which is trained
+to generate reconstructed samples.
+
+For a detailed overview of VQ-VAEs, please refer to the original paper and 
+[this video explanation](https://www.youtube.com/watch?v=VZFVUrYcig0).
+If you need a refresher on VAEs, you can refer to 
+[this book chapter](https://livebook.manning.com/book/deep-learning-with-python-second-edition/chapter-12/).
+VQ-VAEs are one of the main recipes behind [DALL-E](https://openai.com/blog/dall-e/)
+and the idea of a codebook is used in [VQ-GANs](https://arxiv.org/abs/2012.09841).
+
+This example uses references from the
+[official VQ-VAE tutorial](https://github.com/deepmind/sonnet/blob/master/sonnet/examples/vqvae_example.ipynb)
+from DeepMind. To run this example, you will need TensorFlow 2.5 or higher, as well as
+TensorFlow Probability, which can be installed using the command below.
+"""
+
+"""shell
+pip install -q tensorflow-probability
+"""
+
+"""
+## Imports
+"""
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+from tensorflow import keras
+from tensorflow.keras import layers
+import tensorflow_probability as tfp
+import tensorflow as tf
+
+"""
+## `VectorQuantizer` layer
+
+Here, we will implement a custom layer to encapsulate the vector
+quantizer logic, which is the central component of VQ-VAEs.
+Consider an output from the encoder, with shape `(batch_size, height, width, num_channels)`.
+The vector quantizer will first
+flatten this output, only keeping the `num_channels` dimension intact. So, the shape would
+become `(batch_size * height * width, num_channels)`. The rationale behind this is to
+treat the total number of channels as the space for the latent embeddings.
+
+An embedding table is then initialized to learn a codebook. We measure the L2-normalized
+distance between the flattened encoder outputs and code words of this codebook. We take the
+code that yields the minimum distance, and we apply one-hot encoding to achieve quantization.
+This way, the code yielding the minimum distance to the corresponding encoder output is
+mapped as one and the remaining codes are mapped as zeros.
+
+Since the quantization process is not differentiable, we apply a 
+[straight-through estimator](https://www.hassanaskary.com/python/pytorch/deep%20learning/2020/09/19/intuitive-explanation-of-straight-through-estimators.html)
+in between the decoder and the encoder, so that the decoder gradients are directly propagated
+to the encoder. As the encoder and decoder share the same channel space, the hope is that the
+decoder gradients will still be meaningful to the encoder.
+"""
+
+
+class VectorQuantizer(layers.Layer):
+    def __init__(self, num_embeddings, embedding_dim, beta=0.25, **kwargs):
+        super().__init__(**kwargs)
+        self.embedding_dim = embedding_dim
+        self.num_embeddings = num_embeddings
+        self.beta = (
+            beta  # This parameter is best kept between [0.25, 2] as per the paper.
```

Comment:
You can move the comment to the line above to avoid the parens.

Reply:
Oh, I wanted to but it was organized by `black`.